pkg/cri/server/container_create_linux.go

```diff
@@ -166,10 +166,13 @@ func (c *criService) containerSpec(
 
 	// Apply envs from image config first, so that envs from container config
 	// can override them.
-	env := append([]string{}, imageConfig.Env...)
+	env := []string{}
 	for _, e := range config.GetEnvs() {
 		env = append(env, e.GetKey()+"="+e.GetValue())
 	}
+	imageConfigEnv := make([]string, len(imageConfig.Env))
+	copy(imageConfigEnv, imageConfig.Env)
+	specOpts = append(specOpts, oci.WithEnv(imageConfigEnv))
 	specOpts = append(specOpts, oci.WithEnv(env))
```

Comment:
So, the new approach works, because `oci.WithEnv()` is called twice;

- first time it overrides the "defaults" (which removes duplicates)
- second time, the `imageConfigEnv` values are now the new "defaults", so the second call removes duplicates (if present).

This means that every caller of this must jump through the same hoops if they don't want duplicates to be set.

If we don't want duplicates to be set, the alternative would be to update`WithEnv()` to remove duplicates _both_ in defaults, and in "overrides". Something like;

```go
// replaceOrAppendEnvValues returns the defaults with the overrides either
// replaced by env key or appended to the list
func replaceOrAppendEnvValues(defaults, overrides []string) []string {
	cache := make(map[string]int, len(defaults))
	merged := make([]string, 0, len(defaults)+len(overrides))

	for i, e := range defaults {
		parts := strings.SplitN(e, "=", 2)
		merged = append(merged, e)
		cache[parts[0]] = i
	}
	for n, e := range overrides {
		// Values w/o = means they want this env to be removed/unset.
		if !strings.Contains(e, "=") {
			if i, exists := cache[e]; exists {
				merged = append(merged[:i], merged[i+1:]...) // Remove previous value
			}
		} else {
			// Just do a normal set/update
			parts := strings.SplitN(e, "=", 2)
			if i, exists := cache[parts[0]]; exists {
				merged = append(merged[:i], merged[i+1:]...) // Remove previous value
			}
			merged = append(merged, e)
			cache[parts[0]] = n
		}
	}

	return merged
}
```

Reply:
Let me check my code (trying to code while also listening in on a zoom call 😂). I think it may work with a single loop

Reply:
Okay; combined it to a single loop. I think with that code, no changes should be needed in the code, and whenever `SetEnv()` is used, it will;

- Delete values if `SOME_ENV` (without `=`) are included
- Remove old values if `EXISTING_ENV=new_value` is found, and `EXISTING_ENV =new_value` is added at the end
- Same happens if `EXISTING_ENV` occurs multiple times in overrides
- `NEW_ENV=value` is added as usual
- `NEW_ENV` (no value) is ignored if no prior instance of `NEW_ENV` was found.

Reply:
(debugging one mistake I made; hold on)

Reply:
Yeah, we'll probably need two loops after all (updated). But perhaps someone has a good idea how to combine them.

Reply:
sorry for the late reply. I will take a look on this. Will update my comment later~

Reply:
I just spotted this code in Golang's `os/exec` package, which also does de-duplicating (not exactly the same as the requirements we have here, but perhaps nice as example); https://github.com/golang/go/blob/0f64a49460974fa65292d8d5bf6c8a9ceca4b780/src/os/exec/exec.go#L990-L1042

Reply:
This code looks better, should I close this PR?

Reply:
@yeahdongcn I think you can carry it with @thaJeztah suggestion in this pr.

Reply:
Let me do a copy and paste, hold on...

Reply:
done